homeassistant/components/econet/sensor.py

```diff
@@ -1,113 +1,143 @@
+    data = hass.data[DOMAIN][EQUIPMENT][entry.entry_id]
+    equipment = data[EquipmentType.WATER_HEATER].copy()
+    equipment.extend(data[EquipmentType.THERMOSTAT].copy())
 
-    for _equip in all_equipment:
-        for name, attribute in SENSOR_NAMES_TO_ATTRIBUTES.items():
-            if getattr(_equip, attribute, None) is not None:
-                sensors.append(EcoNetSensor(_equip, name))
-        # This is None to start with and all device have it
-        sensors.append(EcoNetSensor(_equip, WIFI_SIGNAL))
+    sensors = []
 
-    for water_heater in equipment[EquipmentType.WATER_HEATER]:
-        # These aren't part of the device and start off as None in pyeconet so always add them
-        sensors.append(EcoNetSensor(water_heater, WATER_USAGE_TODAY))
-        sensors.append(EcoNetSensor(water_heater, POWER_USAGE_TODAY))
+    sensors = [
+        EcoNetSensor(_equip, description)
+        for _equip in equipment
+        for description in SENSOR_TYPES
+        if getattr(_equip, description.key, False) is not False
+    ]
 
     async_add_entities(sensors)
 
 
 class EcoNetSensor(EcoNetEntity, SensorEntity):
     """Define a Econet sensor."""
 
-    def __init__(self, econet_device, device_name):
+    entity_description: SensorEntityDescription
+
+    def __init__(
+        self,
+        econet_device,
+        description: SensorEntityDescription,
+    ) -> None:
         """Initialize."""
         super().__init__(econet_device)
+        self.entity_description = description
         self._econet = econet_device
-        self._device_name = device_name
 
     @property
     def native_value(self):
         """Return sensors state."""
-        value = getattr(self._econet, SENSOR_NAMES_TO_ATTRIBUTES[self._device_name])
+        value = getattr(self._econet, self.entity_description.key)
+        if self.entity_description.name == "power_usage_today":
+            if self._econet.energy_type == ENERGY_KILO_BRITISH_THERMAL_UNIT.upper():
+                value = value * 0.2930710702  # Convert kBtu to kWh
         if isinstance(value, float):
             value = round(value, 2)
         return value
 
     @property
     def native_unit_of_measurement(self):
         """Return the unit of measurement of this entity, if any."""
-        unit_of_measurement = SENSOR_NAMES_TO_UNIT_OF_MEASUREMENT[self._device_name]
-        if self._device_name == POWER_USAGE_TODAY:
-            if self._econet.energy_type == ENERGY_KILO_BRITISH_THERMAL_UNIT.upper():
-                unit_of_measurement = ENERGY_KILO_BRITISH_THERMAL_UNIT
-            else:
-                unit_of_measurement = UnitOfEnergy.KILO_WATT_HOUR
-        return unit_of_measurement
+        return self.entity_description.native_unit_of_measurement
 
     @property
     def name(self) -> str:
         """Return the name of the entity."""
-        return f"{self._econet.device_name}_{self._device_name}"
+        return f"{self._econet.device_name}_{self.entity_description.name}"
```

Comment:
Can this be set inside init?
```suggestion
        self._attr_name = f"{econet_device.device_name}_{description.name}"
```

Reply:
I tried that because that is what ecobee was doing, but it was just getting set to the name of the device I'm guessing because its set in the device class? It was just making a bunch of entities with the device name. eg. Heat Pump Water Heater_1, Heat Pump Water Heater_2, Heat Pump Water Heater_3 ... Or something along those lines. Then when I changed it back it worked.

Reply:
It seems to be due to a conflict with `EcoNetEntity`.
I suggest that you update `EcoNetEntity` to set:
```python
    def __init__(self, econet: Equipment) -> None:
        """Initialize."""
        self._econet = econet
        self._attr_name = econet.device_name
        self._attr_unique_id = f"{econet.device_id}_{econet.device_name}"
```

Reply:
IDK why that didn't fully click. Lol.  That worked. Would we be able to update the name of the entities to not have underscores without breaking current installs? Right now, the power entity shows as `Heat Pump Water Heater_power_usage_today`. Would it make sense to update it, so it's `Heat Pump Water Heater Power Usage Today` or `Heat Pump Water Heater - Power Usage Today`?

Reply:
I suggest that for now:
- you adjust `binary_sensor` the same way as `sensor` to set `attr_name` and `attr_unique_id` inside init (and remove the duplicate `self._econet`)
- you keep the existing name/unique_id convention to avoid a breaking change

Then in a follow-up PR you can address the issue with `name` versus `unique_id`.
Basically, `unique_id` should have been using the `entity_description.key` instead of `entity_description.name`.
Migrating without a breaking change is better done in a specific follow-up PR.

Reply:
Ok sounds good. I'll get around to doing the unique id and name stuff later.